graphql_compiler/schema_transformation/rename_schema.py

```diff
@@ -268,47 +303,128 @@ def enter(self, node, key, parent, path, ancestors):
+            )
         if node.name.value == self.query_type:
             self.in_query_type = False
 
-    def enter_field_definition(self, node, *args):
+    def enter_field_definition(self, node: FieldDefinitionNode, *args):
         """If inside the query type, rename field and add the name pair to reverse_field_map."""
         if self.in_query_type:
             field_name = node.name.value
             new_field_name = self.renamings.get(field_name, field_name)  # Default use original
             if new_field_name == field_name:
                 return None
+            if new_field_name is None:
+                # suppress the type
+                return REMOVE
             else:  # Make copy of node with the changed name, return the copy
                 field_node_with_new_name = get_copy_of_node_with_new_name(node, new_field_name)
                 return field_node_with_new_name
+
+
+class CascadingSuppressionCheckVisitor(Visitor):
+    def __init__(self, renamings: Dict[str, Optional[str]], query_type: str):
+        """Create a visitor to check that suppression does not cause an illegal state.
+
+        Args:
+            renamings: Dict[str, Optional[str]], from original field name to renamed field name. Any
+                       name not in the dict will be unchanged
+            query_type: str, name of the query type (e.g. RootSchemaQuery)
+        """
+        # Note that as field names and type names have been confirmed to match up, any renamed
+        # query type field already has a corresponding renamed type. If no errors, due to either
+        # invalid names or name conflicts, were raised when renaming type, no errors will occur when
+        # renaming query type fields.
+        self.in_query_type = False
+        self.renamings = renamings
+        self.query_type = query_type
+
+    def enter_object_type_definition(self, node: ObjectTypeDefinitionNode, *args):
+        """If the node's name matches the query type, record that we entered the query type."""
+        if node.name.value == self.query_type:
+            self.in_query_type = True
+
+    def leave_object_type_definition(
+        self, node: ObjectTypeDefinitionNode, key, parent, path, ancestors
+    ):
+        """If the node's name matches the query type, record that we left the query type."""
+        if node.name.value == self.query_type:
+            self.in_query_type = False
+
+    def enter_field_definition(self, node: FieldDefinitionNode, *args):
+        """If not at query type, check that no field depends on a type that was suppressed."""
+        if self.in_query_type:
+            return None
+        # at a field of a type that is not the query type
+        field_name = node.name.value
+        node_type = get_ast_with_non_null_and_list_stripped(node.type)
+        if node_type == REMOVE:
+            # then this field depends on a type that was suppressed, which is illegal
+            type_name = args[-1][-1].name.value
```

Comment:
This is an awkward way to get the type that this field belongs to. I haven't found a better way to get this information, but it'd be really useful to figure out how to access both the type in a more natural way.

Also would be really nice to figure out what the type of the field was before the suppression happened, so we could put that information in the error message, but I haven't thought of a clean way to implement that.

Reply:
Are you sure that `*args` is actually intended to be variadic? I think I commented on this above as well — I feel like we may be mis-defining the function signature in a way that just happens to work, but then leads to problems like these.

Reply:
Hm, I just re-read the source code for visitors and it turns out that although other subclasses use `*args`, the superclass `Visitor` actually explicitly lists out all of the arguments so I'll add the arguments explicitly here.